safe_control_gym/controllers/mpc/mpc_acados.py

```diff
@@ -0,0 +1,391 @@
+    print(colored('- To set up the acados python interface, follow the instructions at https://docs.acados.org/python_interface/index.html', 'yellow'))
+    print()
+    exit()
+
+
+class MPC_ACADOS(MPC):
+    '''MPC with full nonlinear model.'''
+
+    def __init__(
+            self,
+            env_func,
+            horizon: int = 5,
+            q_mpc: list = [1],
+            r_mpc: list = [1],
+            warmstart: bool = True,
+            soft_constraints: bool = False,
+            soft_penalty: float = 10000,
+            terminate_run_on_done: bool = True,
+            constraint_tol: float = 1e-6,
+            # runner args
+            # shared/base args
+            output_dir: str = 'results/temp',
+            additional_constraints: list = None,
+            use_gpu: bool = False,
+            seed: int = 0,
+            use_RTI: bool = False,
+            use_lqr_gain_and_terminal_cost: bool = False,
+            **kwargs
+    ):
+        '''Creates task and controller.
+
+        Args:
+            env_func (Callable): function to instantiate task/environment.
+            horizon (int): mpc planning horizon.
+            q_mpc (list): diagonals of state cost weight.
+            r_mpc (list): diagonals of input/action cost weight.
+            warmstart (bool): if to initialize from previous iteration.
+            soft_constraints (bool): Formulate the constraints as soft constraints.
+            terminate_run_on_done (bool): Terminate the run when the environment returns done or not.
+            constraint_tol (float): Tolerance to add the the constraint as sometimes solvers are not exact.
+            output_dir (str): output directory to write logs and results.
+            additional_constraints (list): List of additional constraints
+            use_gpu (bool): False (use cpu) True (use cuda).
+            seed (int): random seed.
+            use_RTI (bool): Real-time iteration for acados.
+            use_lqr_gain_and_terminal_cost (bool): Use LQR ancillary gain and terminal cost for the MPC.
+        '''
+        for k, v in locals().items():
+            if k != 'self' and k != 'kwargs' and '__' not in k:
+                self.__dict__.update({k: v})
+        super().__init__(
+            env_func,
+            horizon=horizon,
+            q_mpc=q_mpc,
+            r_mpc=r_mpc,
+            warmstart=warmstart,
+            soft_constraints=soft_constraints,
+            soft_penalty=soft_penalty,
+            terminate_run_on_done=terminate_run_on_done,
+            constraint_tol=constraint_tol,
+            output_dir=output_dir,
+            additional_constraints=additional_constraints,
+            # compute_initial_guess_method='ipopt',  # use ipopt initial guess by default
+            use_lqr_gain_and_terminal_cost=use_lqr_gain_and_terminal_cost,
+            use_gpu=use_gpu,
+            seed=seed,
+            **kwargs
+        )
+
+        self.x_guess = None
```

Comment:
would these need to be reset? Does it make sense to put these in the `reset` function and then just calling it?

Reply:
I agree putting it into reset is better, and I have updated this.

Reply:
I agree and have put this to the reset.

Reply:
Do we need to call reset in the init function? Will it throw an error somewhere if x_guess is never set?

Reply:
In MPC code, this will only be used for warm-starting, and the none case will be handled [here](https://github.com/MingxuanChe/safe-control-gym/blob/208e844836f3ae487fe2ff85f50dd34a836e0eee/safe_control_gym/controllers/mpc/mpc_acados.py#L316C1-L320C46). I think this should be safe.